.github/workflows/build.yml

```diff
@@ -145,6 +151,122 @@ jobs:
+              if length != 1 then
+                error("unexpected 'oci-layout' document count: " + length)
+              else .[0] end
+              | if .imageLayoutVersion != "1.0.0" then
+                error("unsupported imageLayoutVersion: " + .imageLayoutVersion)
+              else . end
+            ' "$dir/oci-layout" || return "$?"
+            jq -s '
+              if length != 1 then
+                error("unexpected 'index.json' document count: " + length)
+              else .[0] end
+
+              | if .schemaVersion != 2 then
+                error("unsupported schemaVersion: " + .schemaVersion)
+              else . end
+              | if .mediaType != "application/vnd.oci.image.index.v1+json" and .mediaType then # TODO drop the second half of this validation: https://github.com/moby/buildkit/issues/4595
+                error("unsupported index mediaType: " + .mediaType)
+              else . end
+              | if .manifests | length != 1 then
+                error("expected only one manifests entry, not " + (.manifests | length))
+              else . end
+
+              | .manifests[0] |= (
+                if .mediaType != "application/vnd.oci.image.index.v1+json" then
+                  error("unsupported descriptor mediaType: " + .mediaType)
+                else . end
+                # TODO validate .digest somehow (`crane validate`?) - would also be good to validate all descriptors recursively
+                | if .size < 0 then
+                  error("invalid descriptor size: " + .size)
+                else . end
+              )
+            ' "$dir/index.json" || return "$?"
+            local manifest
+            manifest="$dir/blobs/$(jq -r '.manifests[0].digest | sub(":"; "/")' "$dir/index.json")" || return "$?"
+            jq -s '
+              if length != 1 then
+                error("unexpected image index document count: " + length)
+              else .[0] end
+              | if .schemaVersion != 2 then
+                error("unsupported schemaVersion: " + .schemaVersion)
+              else . end
+              | if .mediaType != "application/vnd.oci.image.index.v1+json" then
+                error("unsupported image index mediaType: " + .mediaType)
+              else . end
+
+              # TODO more validation?
+            ' "$manifest" || return "$?"
+          }
+          validate-oci-layout temp
+
+          mkdir signed
+
+          args+=(
+            --mount "type=bind,src=$PWD/temp,dst=/doi-build/unsigned" # TODO this currently assumes normalized_builder == "buildkit" and !should_use_docker_buildx_driver -- we need to factor that in later (although this signs the attestations, not the image, so buildkit/buildx is the only builder whose output we *can* sign right now)
+            --mount "type=bind,src=$PWD/signed,dst=/doi-build/signed"
+
+            docker/image-signer-verifier:0.3.2@sha256:7f5b2da65febf32d88ef6283b36c6af79e05df6e2bb7a7a930d6ea81f58b3757
```

Comment:
Should this use the pinning mechanism?

Reply:
Eventually it probably should, yeah (or go away entirely, if it becomes something `buildkit`/`buildx` does directly) -- for now, this is all really prototypey so it felt more correct to just embed it for now.